Improvement/2121 update the brand to adapt new design

ChengYanJin opened the pull request:
**Component**: ui

**Context**: We want to update our brand theme(light/dark mode) to meet the new design. And will support the theme switch in dev/2.5

**Summary**: 
- update the light/dark theme.json, add the `custom` in order to customise the theme later.
- update the secondary button style and tooltip overlay style
- cherry-pick some bug fix from dev/2.5 ( 3 commits related to volume)
- update the ui-branding configMap

**Acceptance criteria**: 
light mode:
![lightmode](https://user-images.githubusercontent.com/18453133/70817266-8120c400-1dd1-11ea-8acb-4bc2eb6c3d89.png)
![lightmode2](https://user-images.githubusercontent.com/18453133/70817280-85e57800-1dd1-11ea-873a-1930954a2f36.png)


dark mode:
![darkmode](https://user-images.githubusercontent.com/18453133/70788711-c0c7bb80-1d91-11ea-8d41-2b8c84c24f12.png)
![darkmode2](https://user-images.githubusercontent.com/18453133/70788715-c32a1580-1d91-11ea-8b86-f3b0d44e6139.png)

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2121 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ChengYanJin committed:
ui/volume: add refresh in volume information page

ChengYanJin committed:
ui/monitoring: fix the table display bug

When there is no data in the table, text `no data available` is moving from top to down

ChengYanJin committed:
ui/volume: fix the scroll issue of volume list

ChengYanJin committed:
ui/theme: change the button style and tooltip overlay style

- Change the button style in solutionList, createVolume, NodeList and  NodeVolumes
- Change the tooltio overlay style in clusterMonitoring

Refs: #2121

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2121-update-the-brand-to-adapt-new-design` into
`development/2.4`
* [w/2.5/improvement/2121-update-the-brand-to-adapt-new-design](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/2121-update-the-brand-to-adapt-new-design) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

ChengYanJin committed:
ui/assets: add scality branding for light and dark mode

- add logoPath and themes in the config saga
  themes: to store all the themes in ordert to support the theme switch later
  logoPath: to store the scality logo
- update dark/light mode for scality logo

Refs: #2121

ChengYanJin committed:
ui/theme: redefined theme.json with three themes options

light, dark and custom mode(default is light theme)

Refs: #2121

bert-e commented:
# History mismatch



Merge commit #17d79281a6c8119cba4198d54f0b41448db1033b on the integration branch
`w/2.5/improvement/2121-update-the-brand-to-adapt-new-design` is merging a branch which is neither the current
branch `improvement/2121-update-the-brand-to-adapt-new-design` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/2121-update-the-brand-to-adapt-new-design` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #6e4d87e86abcc486618e77c7c97ba4c618d0a1ef on the integration branch
`w/2.5/improvement/2121-update-the-brand-to-adapt-new-design` is merging a branch which is neither the current
branch `improvement/2121-update-the-brand-to-adapt-new-design` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/2121-update-the-brand-to-adapt-new-design` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #7e40bd456820d784fec2eb23da30ae0ad3808e90 on the integration branch
`w/2.5/improvement/2121-update-the-brand-to-adapt-new-design` is merging a branch which is neither the current
branch `improvement/2121-update-the-brand-to-adapt-new-design` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/2121-update-the-brand-to-adapt-new-design` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin committed:
ui/test: update unit test for config.js

Refs: #2121

carlito-scality commented on ui/public/brand/assets/branding-light.svg:
```diff
@@ -0,0 +1,50 @@
+<svg
```
- I think we do not need to add a new `brand[light/dark].svg` because the difference is small. Only the text color is changed. We can just add `class="theme"` to where should be themed in the current `brand.svg` and add CSS to handle it. Thus, we don't need the `logoPath`
- for the "custom" theme, we definitely need the FULL `logoPath`, since the `logo` could be mounted somewhere else than `process.env.PUBLIC_URL`. For now, we don't need to support this yet. @gdemonet any thoughts?

gdemonet replied:
> I think we do not need to add a new `brand[light/dark].svg` because the difference is small. [...] Thus, we don't need the `logoPath`

True, we can add styling directly to the SVG. We could also handle it in JS, FWIW. In the meantime, having a single path to set a logo, be it a default Scality-provided theme or a custom one, seems a reasonable first iteration. I think we can stick to 2 SVGs for now and open an improvement ticket to handle it some time later.

> for the "custom" theme, we definitely need the FULL `logoPath`, since the `logo` could be mounted somewhere else than `process.env.PUBLIC_URL`. For now, we don't need to support this yet. @gdemonet any thoughts?

The only good reason for the logo not being exposed under the same URL as the UI itself is to use an absolute URL (say, from a CDN). IMHO, it's not urgent.
For now, changing a logo would mean:
- adding it as binary data to the `metalk8s-ui/ui-branding` _ConfigMap_
- referencing it from the root of the application (which is what `PUBLIC_URL` represents) in the `theme.json` `"logoPath"`, e.g. `"/brand/custom_logo.svg"`

I'm however now doubting about the _ConfigMap_ approach. It's sufficient for simple theming through `theme.json` values, but as soon as we start adding multiple files, it becomes cumbersome to say the least. As an example, adding sub-directories under `brand` from this single _ConfigMap_ wouldn't be feasible.
Also, the whole idea of sharing this theming configuration between UIs in the cluster means we need a more flexible approach than this single file.

Some solutions come to mind:
- have _another_ static container for branding assets, and let all UIs expose it with their own _Ingress_
    - (+) platform operators can simply edit files on the Bootstrap (including `theme.json`) and the changes are exposed through all the UIs
    - (-) need to backup, propagate to other Bootstrap nodes (maybe even Infra? along with the UI containers...)
- generate the _ConfigMap_**(s)** from static configuration and files using Salt:
    - (+) users still have a single entrypoint
    - (+) this could be used by Solutions, if they want to share theming assets, e.g. for their NavBar icon
    - (-) all UI containers need to be updated when we add a new _ConfigMap_

... :thinking: What if we try both approaches at once? Have a static `nginx`, with its Ingress for e.g. `/brand`, that exposes files from two _ConfigMap_ sources: system defaults and user configuration. As in the current UI containers, system defaults are used as fallback when user-provided values are not available. However, instead of having those stored at build-time, we will need to mount them (this is necessary to add new system defaults dynamically, e.g. for Solution UIs). User-provided configuration can be generated from Salt / Metal UI / whatever.

If you need to upgrade/downgrade Metal / a Solution, change the system defaults _ConfigMap_(s). The user-provided (or generated) _ConfigMap_(s) will be persisted.

carlito-scality commented on ui/src/ducks/config.js:
```diff
@@ -68,11 +75,23 @@ export function updateLanguageAction(language) {
+    // get the default theme from configMap
+    const defaultThemeMode = result.default;
+    yield put(setThemesAction(result.theme));
+    yield put(setLogoPathAction(result.theme[defaultThemeMode].logo_path));
```
no need to set logoPath since it is already in  `theme `

carlito-scality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -384,6 +384,7 @@ const CreateVolume = props => {
                           type="button"
                           onClick={addLabel}
                           data-cy="add-volume-labels-button"
+                          outlined
```
why `outlined`? because for me `outlined` button means `secondary` action and in this case `add volume` is a `primary` action.

ChengYanJin replied:
I just want to meet Use Design's proposal ;) The button from them is like this
![image](https://user-images.githubusercontent.com/18453133/70901310-d4c22600-1ffa-11ea-83e1-d8fd9a210b76.png)

carlito-scality replied:
It should look like this IMO![image](https://user-images.githubusercontent.com/47394132/70902449-18b62a80-1ffd-11ea-8139-ac669745730f.png)

ChengYanJin replied:
This is `add label` though, but I changed other buttons which should be the primary.

carlito-scality commented on ui/public/brand/theme.json:
```diff
@@ -1,17 +1,43 @@
-    "text": "#ffffff",
-    "border": "#ffffff"
-  }
+  "theme":{
```
You should to update the config.json configMap as well

carlito-scality reviewed (changes requested):
you need to update the

ChengYanJin committed:
ui/salt: update theme.json in ui-branding configMap

Refs: #2121

bert-e commented:
# History mismatch



Merge commit #51ba081568b90716b0c61d34ce8e4e69f3b0a2eb on the integration branch
`w/2.5/improvement/2121-update-the-brand-to-adapt-new-design` is merging a branch which is neither the current
branch `improvement/2121-update-the-brand-to-adapt-new-design` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/2121-update-the-brand-to-adapt-new-design` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

carlito-scality reviewed (approved):
LGTM

ChengYanJin commented:
/approve

bert-e commented:
# History mismatch



Merge commit #51ba081568b90716b0c61d34ce8e4e69f3b0a2eb on the integration branch
`w/2.5/improvement/2121-update-the-brand-to-adapt-new-design` is merging a branch which is neither the current
branch `improvement/2121-update-the-brand-to-adapt-new-design` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/2121-update-the-brand-to-adapt-new-design` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/2121-update-the-brand-to-adapt-new-design` with contents from `improvement/2121-update-the-brand-to-adapt-new-design`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/2121-update-the-brand-to-adapt-new-design origin/development/2.5
 $ git merge origin/improvement/2121-update-the-brand-to-adapt-new-design
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/2121-update-the-brand-to-adapt-new-design
```





*The following options are set:* **approve**

ChengYanJin commented:
/approve

bert-e commented:
# History mismatch



Merge commit #a75dabde15d9373ffb5d4db9dbece76ae889e905 on the integration branch
`w/2.5/improvement/2121-update-the-brand-to-adapt-new-design` is merging a branch which is neither the current
branch `improvement/2121-update-the-brand-to-adapt-new-design` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/2121-update-the-brand-to-adapt-new-design` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/2121-update-the-brand-to-adapt-new-design` with contents from `improvement/2121-update-the-brand-to-adapt-new-design`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/2121-update-the-brand-to-adapt-new-design origin/development/2.5
 $ git merge origin/improvement/2121-update-the-brand-to-adapt-new-design
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/2121-update-the-brand-to-adapt-new-design
```





*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #a75dabde15d9373ffb5d4db9dbece76ae889e905 on the integration branch
`w/2.5/improvement/2121-update-the-brand-to-adapt-new-design` is merging a branch which is neither the current
branch `improvement/2121-update-the-brand-to-adapt-new-design` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/2121-update-the-brand-to-adapt-new-design` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/d7c95e9d978cbccf5365f4227fff005626436292) did not succeed in branch w/2.5/improvement/2121-update-the-brand-to-adapt-new-design.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye chengyanjin.
